File: 1-moderate/guess_the_number/guess_the_number.rb
def guess_the_number(input)
  inputs = input.split(" ")
  min = 0
  max = inputs[0].to_i
  guess = ((max - min) / 2.0).ceil
  inputs[1..-1].each do |answer|
    case answer
    when "Higher"
      min = guess + 1
    when "Lower"
      max = guess - 1
    when "Yay!"
      break
    end
    guess = (min + (max - min) / 2.0).ceil
  end
  guess
end

ARGF.each do |line|
  line = line.strip
  next if line.empty?
  puts guess_the_number(line)
end

File: 1-moderate/guess_the_number/guess_the_number.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"strings"
	"strconv"
	"math"
)

func main() {
	file, err := os.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		str := scanner.Text()
		fmt.Println(guessTheNumber(str))
	}
}

func guessTheNumber(input string) string {
	inputs := strings.Split(input, " ")
	var min, max, guess float64
	min = 0.0
	max, _ = strconv.ParseFloat(inputs[0], 64)
	guess = math.Ceil((max - min) / 2.0)
	for _, answer := range inputs[1:] {
		switch answer {
			case "Higher":
				min = guess + 1
			case "Lower":
				max = guess - 1
			case "Yay!":
				break
		}
		guess = math.Ceil(min + (max - min) / 2.0)
	}
	return fmt.Sprintf("%.f", guess)
}